### ISSUE TITLE
Permit user to specify custom format for new date & time insert function

### ISSUE THREAD

**Issue by Gitoffthelawn** [notepad-plus-plus/notepad-plus-plus#10448]
**Description of the Issue**

With the commit 0733e6f2413636f55f2b7a4b6631c4531730ef61, the feature "Insert date and time" was implemented. It inserts the time and then (after a space) the date.

It will be very helpful to allow the user to specify a custom format for this dynamic text insertion.  Supporting a syntax similar to either of the following will work well:

https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Format_Patterns
https://developer.android.com/reference/java/text/SimpleDateFormat

This will also automatically resolve https://github.com/notepad-plus-plus/notepad-plus-plus/issues/10445

**Comment by ArkadiuszMichalski** [comment]
@Gitoffthelawn 
Last commit https://github.com/notepad-plus-plus/notepad-plus-plus/commit/84c1505fab9a0ec4aaadc170daad43342ee95c58 not cover this issue?

![image](https://user-images.githubusercontent.com/2730894/131367102-31dfc170-7e4c-4aaf-a4a4-3fc5dd710bae.png)

**Comment by mere-human** [comment]
There was a [discussion](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10453#issuecomment-907855916) about extending the custom format even further.
We can use this issue for the suggested enhancement or create a new one.

**Comment by Gitoffthelawn** [comment]
@ArkadiuszMichalski I think it largely does.  Just FYI, the commit was was made after this issue was posted.

@mere-human You are welcome to use this issue if you like to discuss further enhancements.

@donho Thank you for adding this functionality!

It looks like that UI needs a little work.  The "Reverse default date time order" checkbox is irrelevant if you are using a custom format.  A way to add clarity is to have 3 radio buttons:

1. System default date and then time
2. System default time and then date
3. Custom format (with an input field)

**Comment by ArkadiuszMichalski** [comment]
@Gitoffthelawn 

Custom is independent command, so `[ ] Reverse...` only applies to `short/long`.

![image](https://user-images.githubusercontent.com/2730894/131417772-32eb908a-37b7-4448-8a12-f05ab723a39e.png)

**Comment by Gitoffthelawn** [comment]
> 
> 
> @Gitoffthelawn
> 
> Custom is independent command, so `[ ] Reverse...` only applies to `short/long`.
> 
> ...
> 
> Alternatively, 2 more commands could be provide for reverse and remove `[ ]` in Preferences:
> 
> ```
> Reverse Date Time (short)
> Reverse Date Time (long)
> ```

Sounds good.  I'm just referring to the settings UI, where the screenshot posted will lead to confusion for many, but improving the UI will resolve the confusion.

**Comment by mere-human** [comment]
> It looks like that UI needs a little work. The "Reverse default date time order" checkbox is irrelevant if you are using a custom format.

Yes, because it is said on the same line that "reverse" is applied only to "short & long formats".
![image](https://user-images.githubusercontent.com/9664141/131458988-090580af-7be5-4503-8611-e637fd1c141f.png)

**Comment by Gitoffthelawn** [comment]
> > It looks like that UI needs a little work. The "Reverse default date time order" checkbox is irrelevant if you are using a custom format.
> 
> Yes, because it is said on the same line that "reverse" is applied only to "short & long formats".

Right.  You'll just have to trust me that the current design is going to cause confusion.  I've been designing human interfaces for decades, and the current design provides a good example of illustrating multiple common UI design errors:

1. The first error is grouping the unrelated settings for multi-instance and the date format together in the same settings page.
1. The second is assuming that the user knows what "short & long formats" means when they encounter this UI.  The use of those terms is quite familiar to some of us, but will be meaningless to many.
1. The third error is compounded by the second error: The current UI design makes it appear as if the custom format definition is somehow impacted by the checkbox directly above it.

The UI also exhibits a few other common design errors, but they are minor in comparison to the above 3 glaring issues.

**Comment by donho** [comment]
@Gitoffthelawn 
Talk is cheap, show me the design.

**Comment by mere-human** [comment]
> 1. The first error is grouping the unrelated settings for multi-instance and the date format together in the same settings page.

I actually suggested moving these settings to a new page. But my suggestion was ignored.



> 2\. The second is assuming that the user knows what "short & long formats" means when they encounter this UI.

But this is how the commands are named in Edit > Insert menu.

**Comment by Gitoffthelawn** [comment]
> @Gitoffthelawn
> Talk is cheap, show me the design.

@donho 
Being rude is generally not a successful strategy to request help from volunteers.  Ask politely and kindly, and I will be happy to accommodate.  

Please note that I already proposed an improved design in a post above.

**Comment by donho** [comment]
@Gitoffthelawn 
My apology. I didn't mean to be rude.

**Comment by donho** [comment]
@ArkadiuszMichalski 
> Alternatively, 2 more commands could be provide for reverse and remove `[ ]` in Preferences:
> 
> ```
> Reverse Date Time (short)
> Reverse Date Time (long)
> ```

I don't like this solution because it makes the commands more complex.

**Comment by alankilborn** [comment]
Capability was added around Notepad++ 8.1.4.
Recommend CLOSURE of this item.
